packages/capacitor-plugin/android/src/main/koltin/com/outsystems/capacitor/plugins/geolocation/GeolocationPlugin.kt

```diff
@@ -0,0 +1,274 @@
+import com.google.gson.Gson
+import com.outsystems.plugins.osgeolocation.controller.OSGLOCController
+import com.outsystems.plugins.osgeolocation.model.OSGLOCException
+import com.outsystems.plugins.osgeolocation.model.OSGLOCLocationOptions
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.cancel
+import kotlinx.coroutines.launch
+
+@CapacitorPlugin(
+    name = "Geolocation",
+    permissions = [Permission(
+        strings = [Manifest.permission.ACCESS_COARSE_LOCATION, Manifest.permission.ACCESS_FINE_LOCATION],
+        alias = GeolocationPlugin.LOCATION_ALIAS
+    ), Permission(
+        strings = [Manifest.permission.ACCESS_COARSE_LOCATION],
+        alias = GeolocationPlugin.COARSE_LOCATION_ALIAS
+    )]
+)
+class GeolocationPlugin : Plugin() {
+
+    private lateinit var controller: OSGLOCController
+    private val gson by lazy { Gson() }
+    private lateinit var coroutineScope: CoroutineScope
+    private val watchingCalls: MutableMap<String, PluginCall> = mutableMapOf()
+
+    companion object {
+        const val LOCATION_ALIAS: String = "location"
+        const val COARSE_LOCATION_ALIAS: String = "coarseLocation"
+    }
+
+    override fun load() {
+        super.load()
+
+        coroutineScope = CoroutineScope(Dispatchers.Main)
+        val activityLauncher = activity.registerForActivityResult(
+            ActivityResultContracts.StartIntentSenderForResult()
+        ) { result ->
+            CoroutineScope(Dispatchers.Main).launch {
+                controller.onResolvableExceptionResult(result.resultCode)
+            }
+        }
+
+        this.controller = OSGLOCController(
+            LocationServices.getFusedLocationProviderClient(context),
+            activityLauncher
+        )
+
+    }
+
+    override fun handleOnDestroy() {
+        super.handleOnDestroy()
+        coroutineScope.cancel()
+    }
+
+    @PluginMethod
+    override fun checkPermissions(call: PluginCall) {
+        if (controller.areLocationServicesEnabled(context)) {
+            super.checkPermissions(call)
+        } else {
+            call.sendError(OSGeolocationErrors.LOCATION_DISABLED)
+        }
+    }
+
+    @PluginMethod
+    override fun requestPermissions(call: PluginCall) {
+        if (controller.areLocationServicesEnabled(context)) {
+            super.requestPermissions(call)
+        } else {
+            call.sendError(OSGeolocationErrors.LOCATION_DISABLED)
+        }
+    }
+
+    /**
+     * Checks location permission state, requesting them if necessary.
+     * If not, calls getPosition to get the device's position
+     */
+    @PluginMethod
+    fun getCurrentPosition(call: PluginCall) {
+        val alias = getAlias(call)
+        if (getPermissionState(alias) != PermissionState.GRANTED) {
+            requestPermissionForAlias(alias, call, "completeCurrentPosition")
+        } else {
+            getPosition(call)
+        }
```

Comment:
This is pretty much identical to the `watchPosition`. Can't we reuse logic to avoid duplications?

Reply:
Refactored, can you check?